### ISSUE TITLE
Filled local queue does not recover when brokers respond to slowly or fail

### ISSUE THREAD

**Issue by webmakersteve** [edenhill/librdkafka#1018]
Description
===========
librdkafka's internal queue fills up when a broker responds too slowly, and does not properly flush once assignments have changed.

How to reproduce
================
We have a 3 node kafka cluster in our QA environment, with a librdkafka producer using `node-rdkafka` producing messages on average at a rate of about 160 per second. The max queue size is currently set to 1000 (which is low, but we had to keep it that low to reproduce the issue with the low volume we're getting in our QA environment).

When using Apache bench to send to the server a total of 20k requests, which translate to 1 message (~96 bytes) each with all 3 brokers up, everything is fine. 

If we shut two of the 3 brokers down, the application continues to work as expected. It connects to that broker and sends it data there.

However, now it is slower (expected). So if we use the same apache bench script, eventually, the queue gets filled. At this point it won't recover from the full queue and continues rejecting all of the http requests for the reason of "Local: Queue full". The application is current polling every one second, and on every produce attempt.

Turn the other two brokers on. The application connects to them as expected. Run the apache bench script again, local queue full is constantly received. It never properly flushes and only restarting the service will make it work again.

Checklist
=========
Please provide the following information:

 - [x] librdkafka version (release number or git tag): d4fd859036be73986a731e0f670819a3a9a5a60f
 - [x] Apache Kafka version: 0.9.0.0
 - [x] librdkafka client configuration:
        'compression.codec': 'none'
        'log.connection.close': false,
        'retry.backoff.ms': 200,
        'message.send.max.retries': 10,
        'socket.keepalive.enable': true,
	'debug': 'queue'
 - [x] Operating system:
Centos 7.2
 - [x] Provide logs (with `debug=..` as necessary) from librdkafka
These are logged using the callback in our application.
```
fault: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.694Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.694Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.695Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.696Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.696Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.706Z ERROR Default: Producer error on service: SINK-01 (channel=Cluster, service=SINK-01, environment=qa, master=false, origin=producer)
    err: {
      "message": "Local: Queue full",
      "code": -184,
      "errno": -184,
      "origin": "kafka"
    }
00:04:09.706Z ERROR Default: Kafka Error: Error: Local: Queue full (channel=Cluster, service=SINK-01, environment=qa, master=false, topic=topic messageSize=1908, failure="Error: Local: Queue full")
00:04:09.753Z  INFO Default: Kafka Send (channel=Cluster, environment=qa, master=false, topic=topic messageSize=1369)
00:04:09.754Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.754Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 1+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.759Z  INFO Default: librdkafka:log:7| broker02:9092/23: Purging bufq with 0 buffers (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.759Z  INFO Default: librdkafka:log:7| broker02:9092/23: Updating 0 buffers on connection reset (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.759Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [0] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.760Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.760Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.760Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.760Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [0] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.760Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.760Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.761Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.761Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [0] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.761Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.761Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.762Z  INFO Default: librdkafka:log:7| broker02:9092/24: topic [1] 0+0 msgs (channel=Cluster, service=SINK-01, environment=qa, master=false)
00:04:09.835Z  INFO Default: librdkafka:log:7| broker02:9092/22: Purging bufq with 0 buffers (channel=Cluster, service=SINK, environment=qa, master=false)
00:04:09.835Z  INFO Default: librdkafka:log:7| broker02:9092/22: Updating 0 buffers on connection reset (channel=Cluster, service=SINK, environment=qa, master=false)
00:04:09.857Z  INFO Default: Kafka Send (channel=Cluster, environment=qa, master=false, topic=topic messageSize=1908)
00:04:09.903Z ERROR Default: Producer error on service: SINK-01 
```
 - [ ] Provide broker log excerpts
Nothing of note.
 - [x] Critical issue

**Comment by edenhill** [comment]
What is your topic replication factor?
When you stop the two other brokers are the partitions migrating to the remaining broker?
What you bring them back up, are partitions migrating back?

If you enable stats, please check where the message are held up: which partition (toppar) and which queues

**Comment by webmakersteve** [comment]
```
Topic:awesome-topic-name       PartitionCount:2        ReplicationFactor:6     Configs:retention.ms=302000000,max.message.bytes=10000
        Topic: awesome-topic-name      Partition: 0    Leader: 34      Replicas: 34,31,32,33,29,30     Isr: 29,33,32,34,31,30
        Topic: awesome-topic-name      Partition: 1    Leader: 29      Replicas: 29,32,33,34,30,31     Isr: 29,33,32,34,31,30
```

I'm not sure that it is migrating to the remaining broker, but I have to assume it is as we have many other services using the Java client and they are recovering correctly. 

I'm going to try to isolate the issue in a way that can be easily reproduced if I can, but it has a lot of variables.

**Comment by edenhill** [comment]
There's been a bunch of recent fixes to the metadata handling, especially around the cases where leaders need to be re-queried.
Is available on master or the soon to be released 0.9.4

**Comment by webmakersteve** [comment]
This is fixed in 0.9.4 as far as I can tell. Great work!

**Comment by edenhill** [comment]
Woohoo! Thanks!
